### ISSUE TITLE
Use MutationObserver to decorate comments in-place

### ISSUE THREAD

**Issue by kfishkin** [kfishkin/butler#3]
This would allow us to remove the kludge of having to press the 'decorate' button in the toolbar.

**Comment by kfishkin** [comment]
Hmm, MutationObserver isn't working for a variety of reasons. I have a working version that detects when you click on the 'Comments' button and polls, but
a) it won't detect sub-comments, and
b) it's a bit of a kludge.

Still might be an improvement, though.

**Comment by AkshatM** [comment]
I think using the MutationObserver API is a sign we're probably overthinking it. :P The API isn't very powerful and requires explicitly working with nodes and nodelists directly - a very low-level approach that shouldn't have to be necessary given Quora's rich HTML semantics and usage. 

Maybe we should attach an onClick event handler to when someone clicks on a subcomment, and simply attach the butler image there. It's a very dumb way to proceed, since the generated subcomments all have unique ids - but I'm guessing Quora's client-side code has some way of uniquely identifying every subcomment before it is loaded.

Alternatively, instead of having a separate button appear on every single comment/subcomment on the popup.js, we could rewrite the extension so that _hovering_ over a comment makes the Butler appear next to it. This has the benefit of being simpler to do - we could attach a global `onHover` event handler to every generic comment class, and use _that_ element's target value to navigate to the adjoining button and handle everything - and being, at least in my opinion, a better user experience. What do you think?

**Comment by AkshatM** [comment]
I'm able to get the mouseover event to fire on any arbitrary subcomment I want without it being a kludge. That's the good news: it means we can capture dynamically generated comments as events without having to resort to MutationObserver.

The bad news is that event bubbling is getting in the way. What I'm trying to do is register the mouseover event only on the outermost tag containing the comment - however, when hovering over the comment, the innermost tags are the ones that are firing their event instead. This is problematic: it means that the events effectively never fires. 

Do you know a good way around this?

**Comment by kfishkin** [comment]
On Sun, Jan 31, 2016 at 11:35 PM, Akshat Mahajan notifications@github.com
wrote:

> I'm able to get the mouseover event to fire on any arbitrary subcomment I
> want without it being a cludge. That's the good news: it means we can
> capture dynamically generated comments as events without having to resort
> to MutationObserver.

hmm - does this include subcomments that aren't there on initial page load?
For example, if you expand a comment and it has subcomments? If so, that's
cool. If not, that's where I stopped as well...

> The bad news is that event bubbling is getting in the way. What I'm trying
> to do is register the mouseover event only on the outermost tag containing
> the comment - however, when hovering over the comment, the innermost tags
> are the ones that are firing their event instead. This is problematic: it
> means that the events effectively never fires.
> 
> Do you know a good way around this?

So it doesn't bubble up to the outermost tags? If so, then I think we're
kinda hosed, we can't stop the stop-bubbling, if you know what I mean. The
only option would be to register on the innermost tags.

**Comment by AkshatM** [comment]
> hmm - does this include subcomments that aren't there on initial page load?
> For example, if you expand a comment and it has subcomments? If so, that's
> cool. If not, that's where I stopped as well...

Yes, that is actually surprisingly straightforward to do. I tackled this problem by listening for events _inside_ the body (which never changes altogether, so I never have to worry about attaching an event handler to a new instance of this), and then simply identifying the target of the event by class name. Something like this:

```
$("body").mouseover(function(event){
    if (event.target.className === "random class name") {
        // do something here
    }
})'
```

works universally across all elements, be they pre-loaded or inserted into the DOM dynamically. The event.target handle is the secret sauce - I'm actually getting a reference to the target _inside_ the DOM, so I can manipulate nearby elements / traverse the DOM / etc. etc. 

This is related to the other problem, however (which isn't exactly event-bubbling, I mischaracterised it). Because I'm no longer listening for specific elements, but merely filtering through all the fired events, this prevents me from specifying that I want all descendants of a tag to respond in the same way. 

In other words, mousing over something like 

```
<div class = "number1">
    <div class = "number2"></div>
</div>
```

will give me _just_ `number2` when I hover over that inner element. If I want to specify that _all_ descendant tags of a tag should do something on mouseover, I need to list out every possible inner tag and write individual handlers for that (since every call to `.parentElement` is different depending on where you are in a triply- or quadruply- nested tag).

This wouldn't be a problem if the comments were shallowly nested, but Quora's comments structure is, quite frankly, insane. They have wrappers within wrappers within wrappers, some of which are only uniquely identified by randomly generated IDs. Writing out handlers for everything is a good way to add needless bloat and run into awful edge cases.

**Comment by AkshatM** [comment]
> So it doesn't bubble up to the outermost tags? If so, then I think we're
> kinda hosed, we can't stop the stop-bubbling, if you know what I mean. The
> only option would be to register on the innermost tags.

If we _do_ register on the innermost tags, though, what should we select? 

Quora has a footer wrapper for every comment that contains the reply, share, downvote and timestamp options. I was thinking this would be ideal, since a simple call to `.parentElement` can get us access to the same space that the Butler image currently occupies - it would mean minimal code rewriting.

The issue is that the space where the footer wrapper can fire an event by itself is such a narrow space (basically, everything to the right of the timestamp in a comment, and nothing above that line height)  that it might be hard for people to figure out how to use the butler. 

What do you think?

**Comment by kfishkin** [comment]
first of all, thanks for diving into this!

second of all, i agree that their comment structure is insane. Their naming
standards are all over the place, too.

but on to the issue at hand - i think jQuery can rescue us here.
If we can trap mouseenter/mousexit at the top 'wrapper' level (I think this
is the div.comment_wrapper), then i think it's all good. Just do this:

on mouseenter -
   see if you have a descendant 'reply' button. if not, you're done.
   if so, then
      if it has a 'butler' icon next to it then
        do nothing - you're done
      else
         add the butler icon, and set up its click and hover handlers

on mouseexit
    see if you have a descendant butler icon. if so, then either kill it or
hide() it, either way seems ok.
